### PR TITLE
k8s cni integration and related netplugin changes/fixes

### PR DESCRIPTION
**Pull request by jojimt**

### CODE REVIEW

File: netmaster/master/endpointGroup.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ package master
 
 import (
 	"errors"
-	"fmt"
+	//"fmt"
```
Comment (jojimt) [inline netmaster/master/endpointGroup.go]: Looks like one commit is missing. Let me fix this.

**Comment by jainvipin** [comment]
@jojimt - can the title of the PR be `k8s cni integration and related netplugin changes/fixes`

**Comment by jojimt** [comment]
Corrected it (what was meant to be description appeared as title…)

**Comment by jainvipin** [comment]
@jojimt: we also need to update the docs with k8 integration, either with this PR or separate one, how to use k8s with contiv, etc.

**Comment by jojimt** [comment]
Yes, that will be part of the next (Vagrant) PR.

File: contivk8s/clients/network.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
+/***
```
Comment (shaleman) [inline contivk8s/clients/network.go]: As i mentioned, can we move contivk8s into mgmtfn/k8splugin hierarchy?

File: netmaster/master/endpointGroup.go [hunk 3]
```diff
@@ -80,13 +80,15 @@ func CreateEndpointGroup(tenantName, networkName, groupName string, epgID int) e
 	}
 
 	// params for docker network
-	subnetCIDR := fmt.Sprintf("%s/%d", nwCfg.SubnetIP, nwCfg.SubnetLen)
+	if GetClusterMode() == "docker" {
```
Comment (shaleman) [inline netmaster/master/endpointGroup.go]: Better way to do this might be to create mgmtfn abstraction here. so that we can move docker/swarm specific things into docker mgmtfn.
We can do this as a subsequent PR too.

File: contivk8s/clients/network.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/netplugin/mgmtfn/cniapi"
+	"io/ioutil"
+	"net"
+	"net/http"
+)
+
+const (
+	nwURL = "http://localhost"
```
Comment (shaleman) [inline contivk8s/clients/network.go]: I'm not sure if we need this since we are using unix sockets

File: contivk8s/k8s_cni.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	ppInfo.IntfName = os.Getenv("CNI_IFNAME")
+	return nil
+}
+
+func addPodToContiv(nc *clients.NWClient, pInfo *cniapi.CNIPodAttr) {
+
+	// Add to contiv network
+	result, err := nc.AddPod(pInfo)
+	if err != nil {
+		log.Fatalf("EP create failed -- %s", err)
```
Comment (shaleman) [inline contivk8s/k8s_cni.go]: How is error communicated back to kubelet? is it based on exit codes?
does log.Fatalf() set the correct exit code?

File: mgmtfn/k8splugin/driver.go [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+	if !ok {
+		return -1, fmt.Errorf("Invalid nw name space: %v", ns)
+	}
+
+	elements := strings.Split(ns, "/")
+	return strconv.Atoi(elements[2])
+}
+
+// setIfAttrs sets the required attributes for the container interface
+func setIfAttrs(ifname, netns, cidr, newname string) error {
```
Comment (shaleman) [inline mgmtfn/k8splugin/driver.go]: Here is how old crtclient code was moving container to namespace without using nester,
https://github.com/contiv/netplugin/blob/k8s-crt-pre-deprecate/crtclient/docker/docker.go#L122

File: mgmtfn/k8splugin/driver.go [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+	resp := epSpec{}
+
+	// Get labels from the kube api server
+	epg, err := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "net-group")
+	if err != nil {
+		log.Errorf("Error getting epg. Err: %v", err)
+		return &resp, err
+	}
+
+	netw, _ := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "network")
```
Comment (shaleman) [inline mgmtfn/k8splugin/driver.go]: error handling missing? :)

File: mgmtfn/cniapi/api.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+package cniapi
+
+// cni api definitions shared between the cni executable and netplugin
+
+// PluginPath is the path to the listen socket directory for netplugin
+const PluginPath = "/run/contiv"
+
+// DriverName is the name of the listen socket for netplugin
+const DriverName = "netplugin"
```
Comment (shaleman) [inline mgmtfn/cniapi/api.go]: Looks like this const is not being used?

File: mgmtfn/k8splugin/cniserver.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/netplugin/mgmtfn/cniapi"
+	"github.com/contiv/netplugin/netplugin/plugin"
+	"github.com/gorilla/mux"
+)
+
+// contivKubeCfgFile holds credentials to access k8s api server
+const (
+	contivKubeCfgFile = "/opt/contiv/config/contiv.json"
```
Comment (shaleman) [inline mgmtfn/k8splugin/cniserver.go]: I'm not too happy about requiring user to create this config file.
But, i don't have an alternative at this point. so, we can keep it for now.

File: mgmtfn/cniapi/api.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
+/***
```
Comment (shaleman) [inline mgmtfn/cniapi/api.go]: as discussed, lets move this under mgmtfn/k8splugin hierarchy too.

**Comment by shaleman** [comment]
LGTM other than the minor comments

File: contivk8s/clients/network.go (hunk 2 above)
Comment (jojimt) [inline contivk8s/clients/network.go]: Yes, will do that.

File: netmaster/master/endpointGroup.go (hunk 3 above)
Comment (jojimt) [inline netmaster/master/endpointGroup.go]: I agree. However, I would rather do it as a separate PR.

File: contivk8s/clients/network.go (hunk 4 above)
Comment (jojimt) [inline contivk8s/clients/network.go]: Actually we do. This is an artifact of the net/http package where it rejects the URL if it did not have a domain name.

File: contivk8s/k8s_cni.go (hunk 5 above)
Comment (jojimt) [inline contivk8s/k8s_cni.go]: kubelet just checks for status 0 and treats everything else as an error without interpreting it.

File: mgmtfn/k8splugin/driver.go (hunk 6 above)
Comment (jojimt) [inline mgmtfn/k8splugin/driver.go]: Yes I see the netns lib. I am not sure it will do everything we need. But I can look at that separately.

File: mgmtfn/k8splugin/driver.go (hunk 7 above)
Comment (jojimt) [inline mgmtfn/k8splugin/driver.go]: Actually, if the first check passes the subsequent ones are guaranteed to pass. I will add a comment to that effect.

File: mgmtfn/cniapi/api.go (hunk 8 above)
Comment (jojimt) [inline mgmtfn/cniapi/api.go]: Right, will remove.

File: mgmtfn/k8splugin/cniserver.go (hunk 9 above)
Comment (jojimt) [inline mgmtfn/k8splugin/cniserver.go]: For the common scenarios, this is created using ansible; so it's not that bad. But, I agree.

File: mgmtfn/cniapi/api.go (hunk 10 above)
Comment (jojimt) [inline mgmtfn/cniapi/api.go]: sure.

File: mgmtfn/k8splugin/cniserver.go (hunk 9 above)
Comment (jainvipin) [inline mgmtfn/k8splugin/cniserver.go]: Is this a security problem?

File: mgmtfn/k8splugin/contivk8s/clients/network.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	c.baseURL = nwURL
+
+	transport := &http.Transport{Dial: unixDial}
+	c.client = &http.Client{Transport: transport}
+
+	return &c
+}
+
+// AddPod adds a pod to contiv using the cni api
+func (c *NWClient) AddPod(podInfo interface{}) (*cniapi.RspAddPod, error) {
```
Comment (jainvipin) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: Unit tests for AddPod and DelPod public functions?

File: mgmtfn/k8splugin/cniserver.go (hunk 9 above)
Comment (jojimt) [inline mgmtfn/k8splugin/cniserver.go]: Well, kube set up uses these certs. Therefore, I don't consider it a security problem specific to contiv. There is also a different way of authenticating that I can look into, separately. However, I don't know that it is any more secure.

File: mgmtfn/k8splugin/cniserver.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+)
+
+// ContivConfig holds information passed via config file during cluster set up
+type ContivConfig struct {
+	K8sAPIServer string `json:"K8S_API_SERVER,omitempty"`
```
Comment (jainvipin) [inline mgmtfn/k8splugin/cniserver.go]: Do we specify which version of API that APIServer interaction is on with?

File: mgmtfn/k8splugin/contivk8s/clients/network.go (hunk 11 above)
Comment (jojimt) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: Because we moved the kubeserver access inside netplugin, these functions will need access to a kube server. I think sys-test is the only option. Will add it after vagrant work is completed.

File: mgmtfn/k8splugin/cniserver.go (hunk 12 above)
Comment (jojimt) [inline mgmtfn/k8splugin/cniserver.go]: Right now it's hard coded to v1 like this:
const (
        nsURL = "/api/v1/namespaces/"
)
We can make it configurable, if necessary.

Comment (jainvipin) [inline mgmtfn/k8splugin/cniserver.go]: Ah... as long it is specified; I don't think there is a need for it to be configurable.

File: mgmtfn/k8splugin/contivk8s/clients/network.go (hunk 11 above)
Comment (jainvipin) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: There should be a way to mock k8s REST handler; for example someone had mentioned an easy way to mock if we made the correct REST calls. Ultimately if you see these functions changing, then making unit tests pays off...

File: mgmtfn/k8splugin/contivk8s/clients/network.go [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+	body := bytes.NewBuffer(buf)
+	url := c.baseURL + cniapi.EPDelURL
+	r, err := c.client.Post(url, "application/json", body)
+	if err != nil {
+		return err
+	}
+	defer r.Body.Close()
+
+	switch {
+	case r.StatusCode == int(404):
```
Comment (jainvipin) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: This code in switch statement looks similar in AddPod and DelPod; shoudl we combine the error handling in a common place?

Comment (jojimt) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: OK, will do.

File: mgmtfn/k8splugin/driver.go [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+		if err != nil {
+			log.Errorf("unable to find link %q. Error %q", ifname, err)
+		}
+		return link, err
+	}
+	return link, err
+}
+
+// nsToPID is a utility that extracts the PID from the netns
+func nsToPID(ns string) (int, error) {
```
Comment (jainvipin) [inline mgmtfn/k8splugin/driver.go]: Perhaps we can use this in mesos plugin too... should this belong in netplugin/utils?

File: mgmtfn/k8splugin/contivk8s/clients/network.go (hunk 11 above)
Comment (jojimt) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: I guess I can do that. Will add it.

File: mgmtfn/k8splugin/driver.go [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+	epReq, err := getEPSpec(&pInfo)
+	if err != nil {
+		log.Errorf("Error getting labels. Err: %v", err)
+		return resp, err
+	}
+
+	err = epCleanUp(epReq)
+	resp.EndpointID = pInfo.InfraContainerID
+	return resp, err
+}
```
Comment (jainvipin) [inline mgmtfn/k8splugin/driver.go]: same comment - no unit tests for any of the functions in this file either...

File: mgmtfn/k8splugin/kubeClient.go [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+	// initiate a get request to the api server
+	podURL := c.baseURL + ns + "/pods/" + name
+	r, err := c.client.Get(podURL)
+	if err != nil {
+		return err
+	}
+
+	defer r.Body.Close()
+	switch {
+	case r.StatusCode == int(404):
```
Comment (jainvipin) [inline mgmtfn/k8splugin/kubeClient.go]: repetitive code - was also there in AddPod and DelPod functions above...

File: mgmtfn/k8splugin/kubeClient.go [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+	// Treat missing metadata as a fatal error
+	if !ok {
+		return fmt.Errorf("metadata not found in podSpec")
+	}
+
+	p := &c.podCache
+	p.setDefaults(ns, name)
+
+	meta := m.(map[string]interface{})
+	l, ok := meta["labels"]
```
Comment (jainvipin) [inline mgmtfn/k8splugin/kubeClient.go]: minor: if statement below this and this statement can be combined

File: netmaster/master/netmaster.go [hunk 18]
```diff
@@ -36,6 +36,13 @@ const (
 	defaultSkyDNSImage  = "skynetservices/skydns:latest"
 )
 
+// Run Time config of netmaster
+type nmRunTimeConf struct {
+	clusterMode string
+}
+
+var masterRTCfg nmRunTimeConf
```
Comment (jainvipin) [inline netmaster/master/netmaster.go]: Should this type of information be added in global runtime structure? This allows for netmaster restartability - @shaleman WDYT?

**Comment by jainvipin** [comment]
@jojimt - great work! thanks!
please identify the changes we need to make before the checkin vs. tests that you want to address as a part of subsequent checkins.

File: netmaster/master/netmaster.go (hunk 18 above)
Comment (jojimt) [inline netmaster/master/netmaster.go]: This is basically environment information. I don't think we should save it in etcd. If the netmaster is restarted, that would be still in the same environment.

Comment (jainvipin) [inline netmaster/master/netmaster.go]: Hmm... got it; is this read always after netmaster start? If yes, we are covered.

Comment (jojimt) [inline netmaster/master/netmaster.go]: It's a command line argument, expected to come from the service env file.

Comment (shaleman) [inline netmaster/master/netmaster.go]: @jainvipin: my thinking is:
(1) Plugin modes (like docker/kubernetes/mesos plugins) that need to be specified before netplugin/netmaster initializes should be command line arguments.
(2) Cluster modes/Fabric modes/control plane modes should be an Intent that can change at runtime. We need little more infra in netplugin/netmaster for this to happen. Expect to work on this in next few PRs.

File: mgmtfn/k8splugin/contivk8s/clients/network.go (hunk 13 above)
Comment (jojimt) [inline mgmtfn/k8splugin/contivk8s/clients/network.go]: After looking at this again, this does not fit as a separate routine as we have different return signatures. It is only 4 lines of code anyway.

File: mgmtfn/k8splugin/driver.go (hunk 14 above)
Comment (jojimt) [inline mgmtfn/k8splugin/driver.go]: As we don't know the exact needs of mesos plugin at this point, I will leave this to @unclejack to refactor and move as appropriate.

File: mgmtfn/k8splugin/driver.go (hunk 15 above)
Comment (jojimt) [inline mgmtfn/k8splugin/driver.go]: I have added UT for the REST interface. This one, seems like it will need to be covered by system test as this module makes multiple external calls.

File: mgmtfn/k8splugin/kubeClient.go (hunk 17 above)
Comment (jojimt) [inline mgmtfn/k8splugin/kubeClient.go]: I have intentionally stayed away from that style because I have seen that it limits the scope of the boolean to just the if. I find that scoping confusing.

**Comment by jainvipin** [comment]
LGTM
